services/QuillLMS/app/models/feedback_history.rb

```diff
@@ -94,11 +95,60 @@ def serializable_hash(options = nil)
     ))
   end
 
+  def serialize_by_activity_session
+   serializable_hash(only: [:session_uid, :start_date, :activity_id, :because_attempts, :but_attempts, :so_attempts, :complete], include: []).symbolize_keys
+  end
+
+  def serialize_by_activity_session_detail
+   serializable_hash(only: [:entry, :feedback_text, :feedback_type, :optimal, :used], include: []).symbolize_keys
+  end
+
   def self.batch_create(param_array)
     param_array.map { |params| create(params) }
   end
 
   private def confirm_prompt_type
     self.prompt_type = DEFAULT_PROMPT_TYPE if prompt_id && !prompt_type
   end
+
+  def self.list_by_activity_session(activity_id: nil, page: 1, page_size: DEFAULT_PAGE_SIZE)
+    query = select(%{
+        feedback_histories.activity_session_uid AS session_uid,
+        MIN(feedback_histories.time) AS start_date,
+        comprehension_prompts.activity_id,
+        COUNT(CASE WHEN comprehension_prompts.conjunction = 'because' THEN 1 END) AS because_attempts,
+        COUNT(CASE WHEN comprehension_prompts.conjunction = 'but' THEN 1 END) AS but_attempts,
+        COUNT(CASE WHEN comprehension_prompts.conjunction = 'so' THEN 1 END) AS so_attempts,
+        (
```

Comment:
The complexity of this `complete` query makes me think we should record completion when it happens.

Reply:
This would significantly complicate the recording of `FeedbackHistory` records.  We'd essentially have to run this conditional, or something like it, scoped to the session each time we insert a new `FeedbackHistory` record in order to determine if the activity is now complete.  I think that a roll-up with logic this complex is sort of at odds with our initial goal with Feedback in general which is that it's mostly state-less.  `complete` requires a lot of state knowledge.

Reply:
Yeah, sorry, I was just thinking aloud here. I should have specified that. I do think the complexity of that query means we should record it somewhere else to make it more queryable, but that's outside of the scope here.